12/README.md

```diff
@@ -0,0 +1,165 @@
+
+* Client library SHOULD provide user an API to register user defined event callbacks  
+
+### Negotiation
+
+![Subscription_2.png](images/Subscription_2.png)
+
+Negotiation phase implies that client executes admin's EventSubscriptionChange command:
+
+* Admin MUST implement EventSubscriptionChange command: 
+
+```ABNF
+EventSubscriptionChange = Tango Command
+EventSubscriptionChange_ARGIN-type  = DevVarStringArr 
+EventSubscriptionChange_ARGIN-value = "[ (upstream device), (attribute), "subscribe", EVENT_TYPE]"
+EventSubscriptionChange_ARGOUT-type = DevLong
+EventSubscriptionChange_ARGOUT-value = "(server library release ver)"
+```
+
+**NOTE**: Tango Server Library v9.3.3 force client to explicitly call ZmqEventSubscriptionChange to use ZMQ based implementation [RFC-13]: 
+ 
+```ABNF
+ZmqEventSubscriptionChange = Tango Command
+ZmqEventSubscriptionChangeArgument-type  = DevVarStringArr 
+ZmqEventSubscriptionChangeArgument-value = "[ upstream device, attribute, action ("subscribe|unsubscribe"), EVENT_TYPE]"
+ZmqEventSubscriptionChangeOutput-type = DevVarLongStringArr
+ZmqEventSubscriptionChangeOutput-value = "[(server library release ver), (upstream device idl ver), (zmq_sub_event_hwm), (rate), (ivl), (zmq release ver); (heartbeat_endpoint), (event_endpoint), n*(alternate_heartbeat_endpoint), n*(alternate_event_endpoint),(heartbeat_channel), (event_channel)]"
+```
+
+* Client MUST execute admin's EventSubscriptionChange command to establish Publisher-Subscriber relationship
+
+* Client SHOULD establish tcp? connection to heartbeat/event endpoints
+
+* Client MAY iterate over alternative heartbeat/event endpoints to establish connection if above has failed
+
+* Client MAY use provided heartbeat/event channels to identify event source 
+
+### Subscription 
+
+* Server MUST send events via corresponding EVENT_CHANNEL
+
+* Server MUST send HEARTBEAT event every 9 seconds
+
+* Client MUST execute EventConfirmSubscription command every 600 seconds to indicate its interest in subscription
+
+```abnf
+EventConfirmSubscription
+```  
+
+### Cancel event subscription
+
+* Client MAY cancel its subscription by executing EventSubscriptionChange on upstream admin 
+
+```abnf
+
+EventSubscriptionChange_ARGIN-value = "[ (upstream device), (attribute), "unsubscribe", EVENT_TYPE]"
+
+```
+
+* Server MUST stop to send events to client whose subscription is canceled
+
+### Fault tolerance
+
+Tango Pub-Sub protocol implies a number of fault tolerance techniques:
+
+- server heartbeat
+- client subscription confirmation
+- events client/server buffer
+
+* Client SHOULD try to re-subscribe once HEARTBEAT event is missing
+
+* Client MAY try to re-subscribe indefinitely
+
+* Server MAY cancel client's subscription if it does not confirm its interest i.e. EventConfirmSubscription was not executed
+
+* Client/Server SHOULD implement (in/out)coming events buffer to prevent overload 
+
+* Client/Server MUST prevent crash in case events overflow
```

Comment:
Clients SHOULD be notified in case of events overflow (e.g.: In the current C++ implementation at least, the client callback is executed with an API_MissedEvents error event when the client library detected that some events got lost. There is a mechanism with some counters associated with each event to help detecting this kind of problems).

Reply:
I prefer what @bourtemb describes than specifying to avoid crash ;-)